server/container/util/src/test/java/org/apache/james/util/scanner/SpamAssassinExtension.java

```diff
@@ -0,0 +1,145 @@
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.testcontainers.containers.GenericContainer;
+
+import com.github.fge.lambdas.Throwing;
+
+public class SpamAssassinExtension implements BeforeAllCallback, AfterAllCallback, ParameterResolver {
+
+    private final GenericContainer<?> spamAssassinContainer;
+    private SpamAssassin spamAssassin;
+
+    public SpamAssassinExtension() {
+        spamAssassinContainer = new GenericContainer<>("dinkel/spamassassin:3.4.0");
+        spamAssassinContainer.waitingFor(new SpamAssassinWaitStrategy());
+    }
+
+    @Override
+    public void beforeAll(ExtensionContext context) throws Exception {
+        spamAssassinContainer.start();
+        spamAssassin = new SpamAssassin(spamAssassinContainer);
+    }
+
+    @Override
+    public void afterAll(ExtensionContext context) throws Exception {
+        spamAssassinContainer.close();
+    }
+
+    @Override
+    public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+        return (parameterContext.getParameter().getType() == SpamAssassin.class);
+    }
+
+    @Override
+    public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext) throws ParameterResolutionException {
+        return spamAssassin;
+    }
+    
+    public static class SpamAssassin {
+        
+        private static final int SPAMASSASSIN_PORT = 783;
+
+        private final String ip;
+        private final int bindingPort;
+        private final GenericContainer<?> spamAssassinContainer;
+
+        private SpamAssassin(GenericContainer<?> spamAssassinContainer) {
+            this.spamAssassinContainer = spamAssassinContainer;
+            this.ip = spamAssassinContainer.getContainerIpAddress();
+            this.bindingPort = spamAssassinContainer.getMappedPort(SPAMASSASSIN_PORT);
+        }
+
+        public String getIp() {
+            return ip;
+        }
+    
+        public int getBindingPort() {
+            return bindingPort;
+        }
+
+        public void train() throws IOException, URISyntaxException {
+            train(Paths.get(ClassLoader.getSystemResource("spamassassin_db/spam").toURI()), TrainingKind.SPAM);
+            train(Paths.get(ClassLoader.getSystemResource("spamassassin_db/ham").toURI()), TrainingKind.HAM);
+        }
+
+        private void train(Path folder, TrainingKind trainingKind) throws URISyntaxException, IOException {
+            try (Stream<Path> paths = Files.walk(folder)) {
+                paths
+                    .filter(Files::isRegularFile)
+                    .map(Path::toFile)
+                    .map(Throwing.function(FileInputStream::new))
+                    .map(readAsPair())
```

Comment:
you should merge readAsPair, closeStream and Pair::getRight in a single method to make code simpler

Reply:
Reworked.